# Проект по SQL


Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Задания**

1. Посчитать, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайть количество обзоров и среднюю оценку;
3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
4. Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Этапы работы:**

1. Импорт библиотек и таблиц.
2. Ответы на вопросы.
3. Общий вывод.



## 1. Импорт библиотек и таблиц.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
#import pandas as pd
#from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
"""
query = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
display(books.head())
display(books.info())
"""

"\nquery = '''\n            SELECT * \n            FROM books\n        '''\nbooks = pd.io.sql.read_sql(query, con = engine)\ndisplay(books.head())\ndisplay(books.info())\n"

In [5]:
"""
query = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
display(authors.head())
display(authors.info())
"""

"\nquery = '''\n            SELECT * \n            FROM authors\n        '''\nauthors = pd.io.sql.read_sql(query, con = engine)\ndisplay(authors.head())\ndisplay(authors.info())\n"

In [6]:
"""
query = '''
            SELECT * 
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
display(publishers.head())
display(publishers.info())
"""

"\nquery = '''\n            SELECT * \n            FROM publishers\n        '''\npublishers = pd.io.sql.read_sql(query, con = engine)\ndisplay(publishers.head())\ndisplay(publishers.info())\n"

In [7]:
"""
query = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
display(ratings.head())
display(ratings.info())
"""

"\nquery = '''\n            SELECT * \n            FROM ratings\n        '''\nratings = pd.io.sql.read_sql(query, con = engine)\ndisplay(ratings.head())\ndisplay(ratings.info())\n"

In [8]:
"""
query = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
display(reviews.head())
display(reviews.info())
"""

"\nquery = '''\n            SELECT * \n            FROM reviews\n        '''\nreviews = pd.io.sql.read_sql(query, con = engine)\ndisplay(reviews.head())\ndisplay(reviews.info())\n"

In [13]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 
    print(f'Таблица - "{table}"')
    display(request.head())
    print(request.info())
    print(f'Таблица - "{table}"')
    print('====================================================================')

Таблица - "books"


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица - "books"
Таблица - "authors"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Таблица - "authors"
Таблица - "publishers"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица - "publishers"
Таблица - "ratings"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица - "ratings"
Таблица - "reviews"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Таблица - "reviews"


## 2. Задания
- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайть количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### 2.1. Посчитать, сколько книг вышло после 1 января 2000 года.

In [13]:
query = '''
            SELECT COUNT(book_id) count_of_books
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(query, con = engine)

count_of_books
0             819

После 1го января 2000 года вышло 819 книг.

### 2.2. Для каждой книги посчитать количество обзоров и среднюю оценку

In [18]:
query = '''
            SELECT b.title, rat.avg_rating, rev.review_count
            FROM books b
            LEFT JOIN (SELECT book_id, AVG(rating) avg_rating FROM ratings GROUP BY book_id) rat ON b.book_id=rat.book_id
            LEFT JOIN (SELECT book_id, COUNT(*) review_count FROM reviews GROUP BY book_id) rev ON b.book_id=rev.book_id
        '''
pd.io.sql.read_sql(query, con = engine)

title  avg_rating  \
0            The Body in the Library (Miss Marple  #3)    4.500000   
1                                            Galápagos    4.500000   
2                             A Tree Grows in Brooklyn    4.250000   
3    Undaunted Courage: The Pioneering First Missio...    4.000000   
4                                          The Prophet    4.285714   
..                                                 ...         ...   
995                                Alice in Wonderland    4.230769   
996          A Woman of Substance (Emma Harte Saga #1)    5.000000   
997                                          Christine    3.428571   
998  The Magicians' Guild (Black Magician Trilogy  #1)    3.500000   
999                           The Plot Against America    3.000000   

     review_count  
0             2.0  
1             2.0  
2             5.0  
3             2.0  
4             4.0  
..            ...  
995           4.0  
996           2.0  
997           3.0  
998           2.0  
999           2.0  

[1000 rows x 3 columns]

Книги имеют по несколько обзоров, средние оценки варьируются.

### 2.3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

In [15]:
query = '''
            SELECT b.publisher_id, p.publisher, COUNT(*) count_books
            FROM books b
            LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY b.publisher_id, p.publisher
            ORDER BY COUNT(*) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count_books
0           212  Penguin Books           42

Лидером по количеству изданных книг толще 50 страниц является издательство Penguin Books, которое выпустило 42 книги.

### 2.4. Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.

In [23]:
"""
query = '''
            SELECT a.author, AVG(ratings)
            FROM (
            SELECT book_id, AVG(rating) ratings
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(*) > 50
            ) rat
            JOIN (SELECT book_id, author_id FROM books) b ON rat.book_id=b.book_id
            JOIN (SELECT author_id, author FROM authors) a ON b.author_id=a.author_id
            GROUP BY a.author
            ORDER BY avg(ratings) DESC

            
        '''
pd.io.sql.read_sql(query, con = engine)
"""

"\nquery = '''\n            SELECT a.author, AVG(ratings)\n            FROM (\n            SELECT book_id, AVG(rating) ratings\n            FROM ratings\n            GROUP BY book_id\n            HAVING COUNT(*) > 50\n            ) rat\n            JOIN (SELECT book_id, author_id FROM books) b ON rat.book_id=b.book_id\n            JOIN (SELECT author_id, author FROM authors) a ON b.author_id=a.author_id\n            GROUP BY a.author\n            ORDER BY avg(ratings) DESC\n\n            \n        '''\npd.io.sql.read_sql(query, con = engine)\n"

Знаменитая Джоан Роулинг в соавторстве с иллюстратором Mary GrandPrе имеет самую высокую среднюю оценку среди авторов.

In [22]:
query = '''
WITH filtered_books AS
    (SELECT b.book_id
     FROM books b
     LEFT JOIN ratings r ON b.book_id = r.book_id
     GROUP BY b.book_id
     HAVING count(DISTINCT r.rating_id) >= 50),

     avg_ratings AS
    (SELECT a.author,
            round(avg(r.rating), 2) AS avg_rating
     FROM books b
     JOIN filtered_books f ON b.book_id = f.book_id
     LEFT JOIN ratings r ON b.book_id = r.book_id
     LEFT JOIN authors a ON b.author_id = a.author_id
     GROUP BY a.author)
     
SELECT *
FROM avg_ratings a
WHERE a.avg_rating =
        (SELECT max(avg_rating)
         FROM avg_ratings)

'''
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

### 2.5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [26]:
query = '''
            SELECT FLOOR (AVG(count_reviews))
            FROM (
            SELECT username, COUNT(*) count_reviews
            FROM reviews
            WHERE username IN
            (
            SELECT username--, COUNT(*)
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 50
            )
            GROUP BY username) rev
        

        '''
pd.io.sql.read_sql(query, con = engine)

floor
0   24.0

Среднее количество обзоров от пользователей, которые поставили более 50 оценок, 24. То есть не каждая оценка сопровождается обзором.

## 3. Выводы:

1. После 1го января 2000 года вышло 819 книг.
2. Книги имеют по несколько обзоров, средние оценки варьируются.
3. Лидером по количеству изданных книг толще 50 страниц является издательство Penguin Books, которое выпустило 42 книги.
4. Знаменитая Джоан Роулинг в соавторстве с иллюстратором Mary GrandPrе имеет самую высокую среднюю оценку среди авторов.
5. Среднее количество обзоров от пользователей, которые поставили более 50 оценок, 24. То есть не каждая оценка сопровождается обзором.